Json list support

bazsi opened the pull request:
json: added support for arrays using the list() type-hints
    
This patch adds support for generating JSON arrays assuming the input is a syslog-ng list.
**example**: `template("$(format-json message=list($MESSAGE))")`

bazsi commented:
This pull request **introduces 1 alert** when merging 87644172f09500374566dde9ec0c2dc9ede6b8f4 into 8f10d8ff4c9a659878e010bf954d0cbfa511b447 - [view on LGTM.com](https://lgtm.com/projects/g/balabit/syslog-ng/rev/pr-5f6d1fdbd1b6806f03c10befb44955bea11a413c)

**new alerts:**

* 1 for Missing return statement

---

*Comment posted by [LGTM.com](https://lgtm.com)*

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build SUCCESS

gaborznagy commented on lib/scanner/list-scanner/list-scanner.h:
```diff
@@ -40,6 +40,7 @@ struct _ListScanner
 
 void list_scanner_input_va(ListScanner *self, const gchar *arg1, ...);
 void list_scanner_input_gstring_array(ListScanner *self, gint argc, GString *argv[]);
+void list_scanner_input_value(ListScanner *self, const gchar *value, gssize value_len);
```
wouldn't help to name it `list_scanner_input_string`?

gaborznagy commented on lib/scanner/list-scanner/tests/test_list_scanner.c:
```diff
@@ -175,6 +175,21 @@ Test(list_scanner, works_with_gstring_input)
+  assert_no_more_tokens();
+
+  /* nonzero terminated */
+  list_scanner_input_value(list_scanner, "foo,bar,baz", 7);
```
Adding more tests to specify behaviour in case of escaped/non-escaped strings?
`value` is assumed to be a list, so to know the rules of a valid list would be good.
It is specified with test `handles_double_quotes`
I.e. 
- `foo ,bar,baz`
- `",foo",bar,baz`

gaborznagy commented on modules/json/tests/test_format_json.c:
```diff
@@ -124,10 +124,13 @@ Test(format_json, test_format_json_with_type_hints)
                          "{\"i\":\"ifoo(\"}");
   assert_template_format("$(format-json b=boolean(TRUE))",
                          "{\"b\":true}");
+  assert_template_format("$(format-json l=list($comma_value))",
```
Would you like to add more test cases than testing input  `value,with,a,comma`, or format-json tests are more like an integration test, while the lower layer (list-scanner) is tested thoroughly?

bazsi commented:
On Fri, Feb 8, 2019 at 10:55 AM Gábor Nagy <notifications@github.com> wrote:

> *@gaborznagy* commented on this pull request.
> ------------------------------
>
> In lib/scanner/list-scanner/list-scanner.h
> <https://github.com/balabit/syslog-ng/pull/2536#discussion_r255005767>:
>
> > @@ -40,6 +40,7 @@ struct _ListScanner
>
>  void list_scanner_input_va(ListScanner *self, const gchar *arg1, ...);
>  void list_scanner_input_gstring_array(ListScanner *self, gint argc, GString *argv[]);
> +void list_scanner_input_value(ListScanner *self, const gchar *value, gssize value_len);
>
> wouldn't help to name it list_scanner_input_string?
>

that's a better name indeed.


> ------------------------------
>
> In lib/scanner/list-scanner/tests/test_list_scanner.c
> <https://github.com/balabit/syslog-ng/pull/2536#discussion_r255007289>:
>
> > @@ -175,6 +175,21 @@ Test(list_scanner, works_with_gstring_input)
>      g_string_free(argv[i], TRUE);
>  }
>
> +Test(list_scanner, works_with_simple_input)
> +{
> +  list_scanner_input_value(list_scanner, "foo,bar,baz", -1);
> +  assert_next_value_is("foo");
> +  assert_next_value_is("bar");
> +  assert_next_value_is("baz");
> +  assert_no_more_tokens();
> +
> +  /* nonzero terminated */
> +  list_scanner_input_value(list_scanner, "foo,bar,baz", 7);
>
> Adding more tests to specify behaviour in case of escaped/non-escaped
> strings?
> value is assumed to be a list, so to know the rules of a valid list would
> be good.
> It is specified with test handles_double_quotes
> I.e.
>
>    - foo ,bar,baz
>    - ",foo",bar,baz
>
>
The list scanner is tested thouroughly in the other testcases, and this one
only covers one variant of how we can specify input. I don't think
repeating all the testcases again would add value.


>
>
> ------------------------------
>
> In modules/json/tests/test_format_json.c
> <https://github.com/balabit/syslog-ng/pull/2536#discussion_r255010049>:
>
> > @@ -124,10 +124,13 @@ Test(format_json, test_format_json_with_type_hints)
>                           "{\"i\":\"ifoo(\"}");
>    assert_template_format("$(format-json b=boolean(TRUE))",
>                           "{\"b\":true}");
> +  assert_template_format("$(format-json l=list($comma_value))",
>
> Would you like to add more test cases than testing input
> value,with,a,comma, or format-json tests are more like an integration
> test, while the lower layer (list-scanner) is tested thoroughly?
>
>
>
yes. list-scanner is pretty well covered with tests at the lower layers.
with this branch it is up to 100% coverage (used to be 94% but I've added a
few more tests, removed some deadcode).

kira-syslogng commented:
Build SUCCESS
